packages/lisk-p2p/src/peer_pool.ts

```diff
@@ -228,17 +236,32 @@ export class PeerPool extends EventEmitter {
 			// TODO ASAP: We need a new type to account for complete P2PPeerInfo which has all possible fields (e.g. P2PDiscoveredPeerInfo) that way we don't need to have all these checks below.
 			peers: this._pickRandomPeers(MAX_PEER_LIST_BATCH_SIZE).map(
 				(peer: Peer) => {
+					const peerDetailedInfo = peer.detailedPeerInfo;
 					const peerInfo = peer.peerInfo;
 
+					if (peerDetailedInfo) {
```

Comment:
Every peer passed to this map callback should always have `peer.detailedPeerInfo` because `_pickRandomPeers` already checked `peer.peerInfo.isDiscoveredPeer` before. Ideally the type system should ensure this. The `_handleGetAllPeersRequest` method should never send back a undiscovered peers.

- Maybe you can change the `_pickRandomPeers` function to filter based on the existence of `peer.detailedPeerInfo` instead of `peer.peerInfo.isDiscoveredPeer`; the two should always we set together.

Reply:
Actually, `_pickRandomPeers` is iterating through Peer object and give out a list of peer objects filtering `isDisovered`, but Typescript doesn't understand if `isDiscoveredPeer` is true then `detailedpeerInfo` will be available. So in the end, I have to check for this condition before using it.